sys/net/gnrc/link_layer/gomach/gomach.c

```diff
@@ -0,0 +1,2222 @@
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#ifndef LOG_LEVEL
+/**
+ * @brief Default log level define
+ */
+#define LOG_LEVEL LOG_WARNING
+#endif
+
+#include "log.h"
+
+/**
+ * @brief  GoMacH thread's PID
+ */
+static kernel_pid_t gomach_pid;
+
+static void _gomach_init(gnrc_netif_t *netif);
+static int _send(gnrc_netif_t *netif, gnrc_pktsnip_t *pkt);
+static gnrc_pktsnip_t *_recv(gnrc_netif_t *netif);
+static void _gomach_msg_handler(gnrc_netif_t *netif, msg_t *msg);
+static void _gomach_event_cb(netdev_t *dev, netdev_event_t event);
+
+static const gnrc_netif_ops_t gomach_ops = {
+    .init = _gomach_init,
+    .send = _send,
+    .recv = _recv,
+    .get = gnrc_netif_get_from_netdev,
+    .set = gnrc_netif_set_from_netdev,
+    .msg_handler = _gomach_msg_handler,
+};
+
+gnrc_netif_t *gnrc_netif_gomach_create(char *stack, int stacksize,
+                                         char priority, char *name,
+                                         netdev_t *dev)
+{
+    return gnrc_netif_create(stack, stacksize, priority, name, dev,
+                              &gomach_ops);
+}
+
+static gnrc_pktsnip_t *_make_netif_hdr(uint8_t *mhr)
+{
+    gnrc_pktsnip_t *snip;
+    uint8_t src[IEEE802154_LONG_ADDRESS_LEN], dst[IEEE802154_LONG_ADDRESS_LEN];
+    int src_len, dst_len;
+    le_uint16_t _pan_tmp;   /* TODO: hand-up PAN IDs to GNRC? */
+
+    dst_len = ieee802154_get_dst(mhr, dst, &_pan_tmp);
+    src_len = ieee802154_get_src(mhr, src, &_pan_tmp);
+    if ((dst_len < 0) || (src_len < 0)) {
+        DEBUG("_make_netif_hdr: unable to get addresses\n");
+        return NULL;
+    }
+    /* allocate space for header */
+    snip = gnrc_netif_hdr_build(src, (size_t)src_len, dst, (size_t)dst_len);
+    if (snip == NULL) {
+        DEBUG("_make_netif_hdr: no space left in packet buffer\n");
+        return NULL;
+    }
+    /* set broadcast flag for broadcast destination */
+    if ((dst_len == 2) && (dst[0] == 0xff) && (dst[1] == 0xff)) {
+        gnrc_netif_hdr_t *hdr = snip->data;
+        hdr->flags |= GNRC_NETIF_HDR_FLAGS_BROADCAST;
+    }
+    return snip;
+}
+
+static gnrc_pktsnip_t *_recv(gnrc_netif_t *netif)
+{
+    netdev_t *dev = netif->dev;
+    netdev_ieee802154_rx_info_t rx_info;
+    netdev_ieee802154_t *state = (netdev_ieee802154_t *)netif->dev;
+    gnrc_pktsnip_t *pkt = NULL;
+    int bytes_expected = dev->driver->recv(dev, NULL, 0, NULL);
+
+    if (bytes_expected > 0) {
+        int nread;
+
+        pkt = gnrc_pktbuf_add(NULL, NULL, bytes_expected, GNRC_NETTYPE_UNDEF);
+        if (pkt == NULL) {
+            DEBUG("_recv_ieee802154: cannot allocate pktsnip.\n");
+            return NULL;
+        }
+        nread = dev->driver->recv(dev, pkt->data, bytes_expected, &rx_info);
+        if (nread <= 0) {
+            gnrc_pktbuf_release(pkt);
+            return NULL;
+        }
+        if (!(state->flags & NETDEV_IEEE802154_RAW)) {
+            gnrc_pktsnip_t *ieee802154_hdr, *netif_hdr;
```

Comment:
If you do the parsing at the latest possible point (i.e. when you hand it up to the network layer) and store the raw packet (i.e. what you've got from `recv` above) instead, the extension of `gnrc_netif_hdr_t` wouldn't be required.

Reply:
you mean we should release the `netif_hdr_t` pktsnip? That will be handled in [here](https://github.com/RIOT-OS/RIOT/pull/5618/files#diff-d4ab863beb293e5d6f5ce98af945e701R256)

Reply:
No, I meant that you shouldn't even create the `netif_hdr` pktsnip, here, but instead starting at the line you reference. You can achieve this by storing `pkt` in the `netif` struct instead of the `netif_hdr` snip.

Reply:
Em, still not know how to do it.. :-/

Reply:
Some how (I am not deep enough into the code to say how right away), a packet lands in the rx queue of the MAC protocol. I say: don't put the already parsed packet, you are creating below in there, but the raw IEEE 802.15.4 packet. This way you don't need to add the sequence number into the `gnrc_netif_hdr_t` struct (which really doesn't belong there, because no packet outside the MAC and link-layer requires it).

Reply:
But, I do need the packet sequence number for some operation in GoMacH. :-/

Reply:
Well the sequence number is in the IEEE 802.15.4 header, right?

Reply:
What I mean is: do not parse the header here at all, just do the following here:

```C
/* ... */
nread = dev->driver->recv(dev, pkt->data, bytes_expected, &rx_info);
if (nread <= 0) {
    gnrc_pktbuf_release(pkt);
    return NULL;
}
DEBUG("_recv_ieee802154: reallocating.\n");
gnrc_pktbuf_realloc_data(pkt, nread);
return pkt;
```

and everything below do only *just before* you dispatch the packet to the next higher layer. This way you don't need to store the sequence number, but can just read it from the IEEE 802.15.4 header (using `ieee802154_get_seq()`) that is still in your packet.

(and yes, the amendment of the `gnrc_netif_hdr_t` struct is a blocking issue for this PR, because it is really not required)